### ISSUE TITLE
Confusion about ground truth

### ISSUE THREAD

**Issue by sanjeevmk** [yhenon/keras-frcnn#149]
I read RCNN, then Fast RCNN and started reading Faster-RCNN. I've been reading it for a week using this repo to clarify things, but its still extremely confusing.
Reading the code here made a lot of things much clearer. However, I'm still very confused about the data preparation steps prior to the training. For both RCNN and Fast RCNN , the region proposals are created before the training and we have GTs for each proposal. Here, since the regions are proposed dynamically, is the GT also created on the fly while the network is training?

There is anchor, ROI and the GT bounding boxes in each image. Is there any other GT that we prepare to train the anchors and ROIs? 

1. I understand that we train the RPN and Fast RCNN network in alternating fashion. Both of them have classification and regression outputs. Are both training with the exact same GT? If they're both training for the same task, why are they separate? If they're training on different tasks, where did the 2 sets of GTs come from, since we had only 1 set of GT to begin with (the image bounding boxes).
2. Image has N objects lets say, so it has N ground truth bounding boxes. When the RPN outputs k (which is much greater than N) bounding boxes, how are the N GT boxes mapped to the k boxes output? That is, how do we know the GT for box1,box2..boxk , since k is not equal to N.
3. We're supposed to feed the input image to the RPN network, nothing else.  However, in the code I see that we're training with values returned from this generator `data_generators.get_anchor_gt` . Why are we inputing anchors instead of images and corresponding boxes?

Sorry if these questions sound naive. I feel that the paper has missed explaining something very basic - detailed input data preparation throughout the entire process. Its all very messy and not clearly explained, it'd be a great help if someone could shed some light on these.. Thanks!

**Comment by Tbarowski** [comment]
> Is there any other GT that we prepare to train the anchors and ROIs?
Yes!

The groundtruth of the RPN is prepared on the fly in training: For each input image you have a fixed number of anchors based on given anchor sizes, anchor ratios, image height/width and the stride (in the implementation as parameters in the config.py). For each of this anchors the data generator checks, whether the anchor has a 70% overlap with a real groundtruth. If the condition is fullfilled, this anchor is a groundtruth anchor so to say. This is done in data_generator.calc_rpn(). There are some more details for bounding boxes with overlap < 30% which are considered negative.

It's written at the beginning of "3.1.2 Loss" in the frcnn paper.

**Comment by sanjeevmk** [comment]
Thanks! We only know the anchor sizes, ratios as these are set in config. But we don't know the true anchor (x,y)-position before we start training. Anchor locations are created during training right? In Fast-RCNN and RCNN, they created proposals using Selective Search which is an _unsupervised_ way of finding class-agnostic candidate regions. Selective Search is replaced by RPN in Faster-RCNN. So RPN should be able to find proposals in some _unsupervised_ manner. But we clearly use some GT here to propose regions. This is the part I don't get mainly..

In the code, we have: 
`X, Y, img_data = next(data_gen_train) 
`where,` data_gen_train = data_generators.get_anchor_gt` .  This X,Y is sent for RPN training.

How do we know anchor GT if anchor x-y positions haven't been predicted yet?

**Comment by HuangBo-Terraloupe** [comment]
@sanjeevmk 
For my understanding, I am not sure it is right.
The anchor GT contains in train_imgs of
data_gen_train = data_generators.get_anchor_gt(train_imgs, classes_count, C, K.image_dim_ordering(), mode='train')

In train_imgs, you have the GT of bounding box, then you can project the gt bounding box to the output layer of RPN. Then you have the groundtruth of region proposals.

**Comment by Tbarowski** [comment]
> Thanks! We only know the anchor sizes, ratios as these are set in config. But we don't know the true anchor (x,y)-position before we start training. Anchor locations are the output of RPN right? In Fast-RCNN and RCNN, they created proposals using Selective Search which is an unsupervised way of finding class-agnostic candidate regions. Selective Search is replaced by RPN in Faster-RCNN. So RPN should be able to find proposals in some unsupervised manner. But we clearly use some GT here to propose regions. This is the part I don't get mainly..

With the anchor sizes and ratios & all the other stuff you now all (x,y) positions and sizes of the anchor. They are strictly defined! Anchors are not the output of the RPN. The output of the RPN is per anchor and contains an objectness score (Does this anchor contain and object) and an ROI which can be interpreted as a finetuning of the anchor ("Net, can you give me a ROI that fits the object slightly better than the anchor, because the anchor is fixed in size and dimensions"). From all these thousands of anchor evaluations the best 300 (or whatever you define) w.r.t. objectness  are picked. These are your proposals that the FRCNN evaluates. In the FRCNN we have a second step of bounding box refinement (Net, can you give me a ROI that is even better than the proposal?)
